### PR TITLE
use crc32c checksum to validate .ji cache files

### PR DESCRIPTION
**Pull request by stevengj**
As discussed in #21154, the intention for a while has been to use a CRC checksum to validate `.ji` files.  This PR implements that by appending the 4-byte checksum to the `.ji` file after it is created.

It validates the checksum only after a cache file passes the other freshness tests, at which point it is about to be reconstituted/loaded, and the additional time for a checksum seems negligible compared to the cost of loading the cache file.

### CODE REVIEW

**Comment by stevengj** [comment]
For example, `using Compat` takes about 0.02 seconds on my laptop (for a fresh cache), while checksumming the `Compat.ji` file takes about 0.0001 seconds.

**Comment by stevengj** [comment]
Now that there is a release-0.6 branch, can this be merged into master?

File: base/util.jl [hunk 1]
```diff
@@ -756,16 +756,22 @@ if is_windows()
 
 end
 
+# compute sizeof correctly for, strings, arrays, and subarrays of bytes
```
Comment (tkelman) [inline base/util.jl]: no comma after `for`

Comment (stevengj) [inline base/util.jl]: fixed

**Comment by stevengj** [comment]
Okay to merge.  @vtjnash or @yuyichao, do you have any thoughts on this?

File: base/loading.jl [hunk 2]
```diff
@@ -772,6 +780,17 @@ function stale_cachefile(modpath::String, cachefile::String)
                 return true
             end
         end
+
+        # finally, verify that the cache file has a valid checksum
+        data = Mmap.mmap(io, Vector{UInt8}, filesize(io), 0)
+        checksum = ntoh(reinterpret(UInt32, data[end-3:end])[1])
```
Comment (yuyichao) [inline base/loading.jl]: I believe this will throw an error now.....

**Review by yuyichao — CHANGES_REQUESTED** [review]

File: base/loading.jl (hunk 2 above)
Comment (vtjnash) [inline base/loading.jl]: (which it should, since this is definitely not aligned)

Comment (stevengj) [inline base/loading.jl]: Why is it definitely not aligned?  `data[end-3:end]` allocates a new array.

Anyway, I can fix it to not use `reinterpret`.

Comment (stevengj) [inline base/loading.jl]:
```jl
checksum = UInt32(data[end]) + UInt32(data[end-1])<<8 + UInt32(data[end-2])<<16 + UInt32(data[end-3])<<24
```
should do it

**Comment by stevengj** [comment]
Fixed and rebased.

**Comment by stevengj** [comment]
@yuyichao, okay now?

**Review by yuyichao — APPROVED** [review]

**Comment by stevengj** [comment]
Thanks, will merge in the next day or two if there are no further objections.

**Pull request merged**